Index: Settings.cs
===================================================================
--- Settings.cs	(revision 2413)
+++ Settings.cs	(working copy)
@@ -47,9 +47,6 @@
         /// <summary>Beta grid login server</summary>
         public const string ADITI_LOGIN_SERVER = "https://login.aditi.lindenlab.com/cgi-bin/login.cgi";
 
-        /// <summary>The relative directory where external resources are kept</summary>
-        public const string RESOURCE_DIR = "openmetaverse_data";
-
         /// <summary>Login server to connect to</summary>
         public string LOGIN_SERVER = AGNI_LOGIN_SERVER;
 
@@ -263,8 +260,11 @@
         /// </summary>
         public bool USE_TEXTURE_CACHE = false;
 
+        /// <summary>The relative directory where external resources are kept</summary>
+        public static string RESOURCE_DIR = "openmetaverse_data";
+
         /// <summary>Path to store cached texture data</summary>
-        public string TEXTURE_CACHE_DIR = RESOURCE_DIR + "/cache";
+        public string TEXTURE_CACHE_DIR =  "openmetaverse_data/cache";
 
         /// <summary>Maximum size cached files are allowed to take on disk (bytes)</summary>
         public long TEXTURE_CACHE_MAX_SIZE = 1024 * 1024 * 1024; // 1GB
@@ -333,5 +333,10 @@
         }
 
         #endregion
+
+	public void setresourcedir(string dir)
+	{
+		RESOURCE_DIR=dir;
+	}
     }
 }
